### PR TITLE
doc: extensions: add kconfig search extension

### PR DESCRIPTION
**Pull request by gmarull**
# Introduction

This PR adds a new Sphinx extension to handle Kconfig documentation. The way this extension works differs significantly from the current approach, mainly due to technical limitations that will be detailed next. As of today, a bunch of RST pages are generated (one per option, so ~10K) by a script (`gen_kconfig_rest.py`). These pages get parsed and rendered by Sphinx when the documentation is built. It turns out that adding ~10K individual pages to Sphinx slows things down, to a point we had to create the `KCONFIG_TURBO_MODE` or `html-fast` targets. In practice, this means that any viable solution needs to stay away from individual pages. In fact, the way bindings docs work suffer from the same issue, just that we still have less of them.

Technical aspects of the solution proposed here:

- It is a native Sphinx extension (no more CMake hackery required)
- It uses `kconfiglib` to build a JSON database for all Kconfig symbols
- Introduces a new Sphinx language domain for Kconfig. The domain provides a Sphinx directive (`.. kconfig:search::`) that can be used to insert a Kconfig search tool on any RST page. When building HTML docs, that directive gets replaced by an actual search form. The `:kconfig:option:` role is now available to reference options (goes to Kconfig search form pre-selecting the option). Thanks to using a regular Sphinx domain, Kconfig options can also be referenced in other projects using Intersphinx.
- A client-side Javascript library takes care of the rest: fetches the database and, based on user input, it renders Kconfig content dynamically. The library also handles URL hash, that is, if the page holding the Kconfig form is queried with `#KCONFIG_OPTION_TO_DISPLAY` it will just display it. This allows for _permalinks_, e.g., [CONFIG_SPI](https://teslabs.github.io/zephyr-kconfig-test/kconfig.html#CONFIG_SPI).

Pros:
- A native Sphinx extension
- Only a single page is needed: this means docs will be faster by default
- We can quickly search Kconfig options (it's faster than Sphinx search!)

Cons:
- Kconfig options don't have a page now, so search engines will not index them.

# Demo

https://teslabs.github.io/zephyr-kconfig-test/kconfig.html

image

### CODE REVIEW

**Comment by nashif** [comment]
> * Kconfig options don't have a page now, so search engines will not index them.

can we just list them somewhere without any individual rst pages for search engines to index with instructions to use the search capability for more details?

**Comment by gmarull** [comment]
> > * Kconfig options don't have a page now, so search engines will not index them.
> 
> can we just list them somewhere without any individual rst pages for search engines to index with instructions to use the search capability for more details?

Maybe by rendering the whole database will help if crawler is detected: https://developers.google.com/search/docs/advanced/javascript/dynamic-rendering

**Comment by mbolivar-nordic** [comment]
> Maybe by rendering the whole database will help if crawler is detected: https://developers.google.com/search/docs/advanced/javascript/dynamic-rendering

that kind of sounds like an `html-slow` ;)

**Comment by mbolivar-nordic** [comment]
+1 assuming the issue I pointed out is fixed

**Comment by de-nordic** [comment]
Would it be possible to add some html param that would allow to capture search list? This would be useful in case when you need to inform somebody on list of Kconfig options that are dedicated to specific subsystem, for example:
`https://teslabs.github.io/zephyr-kconfig-test/kconfig.html?seach=CONFIG_FS_LITTLEFS_`
would redirect user to page with the `CONFIG_FS_LITTLEFS_` already placed in search bar and list narrowed to Kconfig options beginning with `CONFIG_FS_LITTLEFS_`.

**Comment by tejlmand** [comment]
This looks great.

Some observations.

Old style will print all help texts (just like menuconfig),  meaning very often settings will not duplicate the help.
For example like here: [CONFIG_BT_CTLR_ADV_PERIODIC](https://docs.zephyrproject.org/latest/reference/kconfig/CONFIG_BT_CTLR_ADV_PERIODIC.html#help)

But with new approach, the duplicated setting gets its own description, but without help, like here: https://teslabs.github.io/zephyr-kconfig-test/kconfig.html#CONFIG_BT_CTLR_ADV_PERIODIC

But we should be careful not to start duplicating help texts, cause that will mess up `menuconfig`, for example as it can be seen here:
image

So at the current moment I don't have a good solution to this, and I don't think we should start duplicating help texts (instead we should clean up those that are duplicated today)

Also, due to supporting optional Zephyr modules, we source symbols with an `if 0` causing the symbols to be unconfigurable, while still ensuring Kconfig tree works even when the module is missing.

However, this also results in duplicated symbols (both in old docs and in new):
https://teslabs.github.io/zephyr-kconfig-test/kconfig.html#CONFIG_BUILD_WITH_TFM

But I notice, the location is identical:
image

Is it possible to remove the entry, if the location field is identical to the location field of similar entry ?

**Comment by tejlmand** [comment]
Should we consider to make `prompt` more distinct from the `help` text  ?

**Comment by gmarull** [comment]
> Is it possible to remove the entry, if the location field is identical to the location field of similar entry ?

Yes, it's possible.

> Should we consider to make `prompt` more distinct from the `help` text ?

Maybe render it in **bold** or *italics*?

**Comment by mbolivar-nordic** [comment]
This is definitely a massive improvement. As discussed at dev-review, I remain concerned about the way defaults are displayed. I think it should be clearer to the user that a system-wide default is not the same as what may appear on their target.

**Comment by gmarull** [comment]
@mbolivar-nordic added support for _other_ defaults in latest version

https://user-images.githubusercontent.com/25011557/154976100-011e542b-b901-4af3-94ef-c3e54f9b9e53.mp4

**Review by mbolivar-nordic — DISMISSED** [review]
Excellent

**Comment by tejlmand** [comment]
> > Is it possible to remove the entry, if the location field is identical to the location field of similar entry ?
>
> Yes, it's possible.

Great :+1: 

> > Should we consider to make prompt more distinct from the help text ?
>
> Maybe render it in bold or italics?

Both are fine with me, but I see you already changed to *italics* , so let's keep it with that :+1:

Great work.

**Review by henrikbrixandersen — DISMISSED** [review]

File: CODEOWNERS [hunk 1]
```diff
@@ -611,7 +611,7 @@
 /lib/libc/arcmwdt/                        @abrodkin @ruuddw @evgeniy-paltsev
 /misc/                                    @tejlmand
 /modules/                                 @nashif
-/modules/canopennode/                     @henrikbrixandersen
+/modules/CANopenNode/                     @henrikbrixandersen
```
[the remark this answers is not in the document]
Comment (gmarull) [reply CODEOWNERS:614]: The reason I did this change is that I changed the way to handle `kconfig-ext` cases. canopennode seems to be the only case using a folder name in `modules` != to what is defined in `module.yml`: https://github.com/zephyrproject-rtos/canopennode/blob/zephyr/zephyr/module.yml#L1. Before, I was relying on folder names in `modules/`, which require name escaping (already done by west APIs). I think that the `kconfig-ext` case is not particularly well resolved...

**Comment by gmarull** [comment]
@mbolivar-nordic @henrikbrixandersen @carlescufi @nashif should be ready now

**Review by henrikbrixandersen — COMMENTED** [review]
The documentation changes (and the entire concept) looks fine. Thanks!

I know too little of Sphinx to feel comfortable in reviewing the actual Kconfig extension, though.

**Review by mbolivar-nordic — APPROVED** [review]
I'm giving a +1 for the feature without having reviewed in detail because I'm confident @gmarull  will fix any issues that might get reported after this is merged. Thanks for this.

**Review by nashif — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]

**Pull request merged**